ts/src/profiler.ts

```diff
@@ -13,47 +13,335 @@
+    }
+    if (!this.config.disableHeap) {
+      this.profileTypes.push(ProfileTypes.Heap);
+      this.heapProfiler = new HeapProfiler(
+          this.config.heapIntervalBytes, this.config.heapMaxStackDepth);
+    }
+  }
+
+  /**
+   * Starts an endless loop to poll profiler server for instructions, and
+   * collects and uploads profiles as requested.
+   * If there is a problem when collecting a profile or uploading a profile to
+   * profiler server, this problem will be logged at the error level and
+   * otherwise ignored.
+   * If there is a problem polling profiler server for instructions
+   * on the type of profile to be collected, this problem will be logged at the
+   * error level and getting profile type will be retried.
+   */
+  start() {
+    this.runLoop();
+  }
+
+  /**
+   * Endlessly polls the profiler server for instructions, and collects and
+   * uploads profiles as requested.
+   */
+  async runLoop(): Promise<void> {
+    const delayMillis = await this.collectProfile();
+
+    // Schedule the next profile.
+    setTimeout(this.runLoop.bind(this), delayMillis).unref();
+  }
+
+  /**
+   * Waits for profiler server to tell it to collect a profile, then collects
+   * a profile and uploads it.
+   *
+   * @return - time, in ms, to wait before asking profiler server again about
+   * collecting another profile.
+   *
+   * TODO: implement backoff and retry. When error encountered in
+   * requestProfile() should be retried when response indicates this request
+   * should be retried or with exponential backoff (up to one hour) if the
+   * response does not indicate when to retry this request.
+   */
+  async collectProfile(): Promise<number> {
+    let prof: RequestProfile;
+    try {
+      prof = await this.requestProfile();
+    } catch (err) {
+      this.logger.error(
+          `Error requesting profile type to be collected: ${err}`);
+      return this.config.backoffMillis;
+    }
+    try {
+      await this.profileAndUpload(prof);
+    } catch (err) {
+      this.logger.error(`Error collecting and uploading profile: ${err}`);
+    }
+    return 0;
+  }
+
+  /**
+   * Talks to profiler server, which hangs until server indicates
+   * job should be profiled and then indicates what type of profile should
+   * be collected.
+   *
+   * If any problem is encountered, an error will be thrown.
+   *
+   * @return - a RequestProfile specifying which type of profile should be
+   * collected and other information needed to collect and upload a profile of
+   * the specified type.
+   *
+   * TODO (issue #28): right now, this call could hang for up to an hour when
+   * this method is the only thing on the event loop, keeping the program open
+   * even when all work is done. Should expose the ability to cancel the http
+   * request made here, and then determine when to cancel this request.
+   *
+   * Public to allow for testing.
+   */
+  async requestProfile(): Promise<RequestProfile> {
```

Comment:
Call it `createProfile` to match the API call name.

Reply:
Done.